utils/vector.hpp

```diff
@@ -0,0 +1,281 @@
+    
+    //Copy the values of the existing array to new array
+    for(uint32_t i = 0; i < size; i++)
+    {
+        newArray[i] = array[i];
+    }
+    
+    //Change the data of the vector
+    delete[]   array;
+    array      = newArray;
+    capacity   = size;
+    this->size = size;
+}
+
+template<class T>
+void Vector<T>::resize(uint32_t size, const T& fill)
+{
+    uint32_t oldSize = this->size;
+    resize(size);
+    for (uint32_t i = oldSize; i < size; i++)
+    {
+        array[i] = fill;
+    }
+}
+
+template<class T>
+void Vector<T>::reserve(unsigned int newCap)
+{
+    if(capacity < newCap)
+    {
+        //Keep multiplying capacity by 2 until it's larger than the new capacity
+        while(capacity < newCap)
+            capacity *= 2;
+        
+        //Reallocate the array
+        T* newArray = new T[capacity];
+        
+        //Copy the values of the existing array to new array
+        for(unsigned int i = 0; i < size; i++)
+        {
+            newArray[i] = array[i];
+        }
+        
+        //Delete old array and assign new array to vector
+        delete[] array;
+        array = newArray;
+    }
+}
+
+/* Modifiers */
+template<class T>
+void Vector<T>::push_back(const T& value)
+{
+    if(capacity == size)
+    {
+        reserve(capacity * 2);
+    }
+    array[size] = value;
+    size++;
+}
+
+template<class T>
+void Vector<T>::pop_back()
+{
+    if(size > 0)
+    {
+        size--;
+    }
+}
+
+template<class T>
+void Vector<T>::clear()
+{
+    size = 0;
+    capacity = 0;
+    delete[] array;
+    array = nullptr;
+}
+
+/* Operator Overloads */
+template<class T>
+Vector<T>& Vector<T>::operator=(const Vector& vec)
+{
+    //Delete old array
+    delete[] array;
+    
+    //Resize array to fit new elements
+    resize(vec.size);
+    
+    //Deep copy elements from other vector
+    for (unsigned int i = 0; i < size; i++)
+    {
+        array[i] = vec.array[i];
+    }
+    
+    return *this;
+}
+
+template<class T>
+Vector<T>& Vector<T>::operator=(Vector&& vec)
+{
+    size      = vec.size;
+    array     = vec.array;
+    capacity  = vec.capacity;
+    vec.array = nullptr;
+    return *this;
+}
+
+template <class T>
+T& Vector<T>::operator[](unsigned int index)
+{
+    return array[index];
+}
+
+/* Iterators */
+template <class T>
+T* Vector<T>::begin()
+{
+    return &array[0];
+}
+
+template <class T>
+T* Vector<T>::end()
+{
+    return &array[size];
+}
+
+template <class T>
+const T* Vector<T>::begin() const
+{
+    return static_cast<const T*>(array[0]);
+}
+
+template <class T>
+const T* Vector<T>::end() const
+{
+    return static_cast<const T*>(array[size]);
+}
+
+template <class T>
+void Vector<T>::erase(uint32_t position)
+{
+        for(uint32_t i = position; i < size; i++)
+        {
+            array[i] = array [i + 1];
+        }
+        size--;
+}
+
+template<class T>
+void Vector<T>::insert(uint32_t position, const T& value)
+{
+    
+    if (size >= capacity)
```

Comment:
Size will never be greater than capacity
```suggestion
    if (size > capacity)
```